pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -910,13 +929,19 @@ func (r *RootSyncReconciler) mutationsFor(ctx context.Context, rs *v1beta1.RootS
 				container.Env = append(container.Env, containerEnvs[container.Name]...)
 				mutateContainerResource(&container, rs.Spec.Override)
 			case reconcilermanager.HydrationController:
-				container.Env = append(container.Env, containerEnvs[container.Name]...)
-				if rs.Spec.SafeOverride().EnableShellInRendering == nil || !*rs.Spec.SafeOverride().EnableShellInRendering {
-					container.Image = strings.ReplaceAll(container.Image, reconcilermanager.HydrationControllerWithShell, reconcilermanager.HydrationController)
+				if !enableRendering(rs.GetAnnotations()) {
+					// if the sync source does not require rendering, omit the hydration controller
+					// this minimizes the resource footprint of the reconciler
+					addContainer = false
 				} else {
-					container.Image = strings.ReplaceAll(container.Image, reconcilermanager.HydrationController+":", reconcilermanager.HydrationControllerWithShell+":")
+					container.Env = append(container.Env, containerEnvs[container.Name]...)
+					if rs.Spec.SafeOverride().EnableShellInRendering == nil || !*rs.Spec.SafeOverride().EnableShellInRendering {
+						container.Image = strings.ReplaceAll(container.Image, reconcilermanager.HydrationControllerWithShell, reconcilermanager.HydrationController)
+					} else {
+						container.Image = strings.ReplaceAll(container.Image, reconcilermanager.HydrationController+":", reconcilermanager.HydrationControllerWithShell+":")
+					}
```

Comment:
nit: This adds another layer to the already deeply nested conditionals. Can we improve it to make it more readable? For example, maybe the shell part can be wrapped as a function.

Reply:
fair nit, updated